### ISSUE TITLE
command line SCP working but go-scp is showing timeout error 

### ISSUE THREAD

**Issue by gurdevrana** [bramvdbogaerde/go-scp#57]
I am using go-scp and trying to copy to solarwinds server(windows server) and getting waiting timed out error whereas I tried command line scp it works fine.  

Also I found out that after removing -q option in line 'err := a.Session.Run(fmt.Sprintf("%s -qt %q", a.RemoteBinary, remotePat)'
in CopyPassThru function, there was no waiting timed out error but the file was empty at the remote server

**Comment by bramvdbogaerde** [comment]
Thanks for reporting this issue. I don't have a windows server myself, so I am unable to reproduce this issue. 

The SCP protocol (especially the options on the remote binary) are not exactly standard, and it is very much possible that something is missing in this implementation to communicate correctly with a SolarWinds server.

There were some blocking and error response parsing issues in the past that were fixed in PR #56 , which might also fix your problems.

I will close this issue for now, as I lack the necessary equipment to do experiments myself. Any contributions with regards to insights into this problem are, of course, very welcome!

**Comment by gurdevrana** [comment]
I think there was some issue with synchronizing between the two go routine. I was able to resolve my issue by doing the following changes i.e running session.run in the main thread rather than goroutine :-
func (a *Client) CopyPassThru(r io.Reader, remotePath string, permissions string, size int64, passThru PassThru) error {
	stdout, err := a.Session.StdoutPipe()
	if err != nil {
		return err
	}

	if passThru != nil {
		r = passThru(r, size)
	}

	filename := path.Base(remotePath)

	//wg := sync.WaitGroup{}
	//wg.Add(1)

	errCh := make(chan error, 1)

	go func() {
		//defer wg.Done()
		w, err := a.Session.StdinPipe()
		if err != nil {
			errCh <- err
			return
		}

		defer w.Close()

		_, err = fmt.Fprintln(w, "C"+permissions, size, filename)
		if err != nil {
			errCh <- err
			return
		}

		if err = checkResponse(stdout); err != nil {
			errCh <- err
			return
		}

		_, err = io.Copy(w, r)
		if err != nil {
			errCh <- err
			return
		}

		_, err = fmt.Fprint(w, "\x00")
		if err != nil {
			errCh <- err
			return
		}

		if err = checkResponse(stdout); err != nil {
			errCh <- err
			return
		}
	}()


	err = a.Session.Run(fmt.Sprintf("%s -t %s", a.RemoteBinary, remotePath))
	if err != nil {
		return errors.New("Session failed when upload files")
	}

	//if waitTimeout(&wg, a.Timeout) {
	//	return errors.New("timeout when upload files")
	//}

	close(errCh)
	for err := range errCh {
		if err != nil {
			return err
		}
	}
	return nil
}

**Comment by bramvdbogaerde** [comment]
Good catch! But I think that there are two more problems that would require our attention then:

1) Running the session in the main thread breaks the time-out functionality (you also commented in the waitTimeoutFunction), which was a feature that was requested (and implemented) some time ago. I think it would be useful to keep this feature around.
2) A potentially fatal race condition could occur between the main thread and the Go routine responsible for copying. The reason for this is that StdinPipe() is called in the Go routine rather than in the main thread, which could lead to race conditions where Run is called before calling StdinPipe(), which can lead to errors such as "StdinPipe already set". AFAIK, this could already happen in the current implementation in theory (see issue #39), but it should be easy enough to fix. 

However, I think you are really close to a working solution, and would welcome your contribution!

**Comment by gurdevrana** [comment]
I agree with your concerns. how about we move StdinPipe() to the main thread which I believe will solve problem 2. And for problem 1, we can have a channel like startWriteCh such that before writing it will wait for the signal from session.Run()
So, the final code will look like :- 

`func (a *Client) CopyPassThru(r io.Reader, remotePath string, permissions string, size int64, passThru PassThru) error {
	stdout, err := a.Session.StdoutPipe()
	if err != nil {
		return err
	}

	if passThru != nil {
		r = passThru(r, size)
	}

	filename := path.Base(remotePath)

	wg := sync.WaitGroup{}
	wg.Add(2)

	errCh := make(chan error, 2)
        startWriteCh := make(chan bool)
 	w, err := a.Session.StdinPipe()
	if err != nil {
		return err
	}	
	go func() {
		defer wg.Done()
		defer w.Close()
		<-startWriteCh
		_, err = fmt.Fprintln(w, "C"+permissions, size, filename)
		if err != nil {
			errCh <- err
			return
		}

		if err = checkResponse(stdout); err != nil {
			errCh <- err
			return
		}

		_, err = io.Copy(w, r)
		if err != nil {
			errCh <- err
			return
		}

		_, err = fmt.Fprint(w, "\x00")
		if err != nil {
			errCh <- err
			return
		}

		if err = checkResponse(stdout); err != nil {
			errCh <- err
			return
		}
	}()

   go func(){
        defer wg.Done()
		startWriteCh <- true
		err = a.Session.Run(fmt.Sprintf("%s -t %s", a.RemoteBinary, remotePath))
		if err != nil {
			errCh <- err
			return
		}
	}()

	if waitTimeout(&wg, a.Timeout) {
		return errors.New("timeout when upload files")
	}

	close(errCh)
	for err := range errCh {
		if err != nil {
			return err
		}
	}
	return nil
}`
